PeerioTechnologies/peerio-icebear
# chore: introduce typescript

#### Relevant info and issue/PR links

https://app.clubhouse.io/peerio/story/11183/sdk-convert-to-typescript

Upgrades icebear to Babel 7 -- required for `@babel/preset-typescript` -- and introduces the capacity to incrementally rewrite the sdk. One leaf file has been switched over to TS as a test.

TypeScipt `strict` mode is enabled, with `noImplicitAny` disabled.
 
Tests don't work yet -- they use their own `.babelrc` that I haven't switched over. (I think it could probably be changed to just be shared with the root `.babelrc`?)

Use the temporary npm script `test-build` to invoke Babel.

---

To do:
- [ ] Tests (and desktop, and maybe mobile?) use `babel-root-import` (which is just a year-old, deprecated version of [`babel-plugin-root-import`](https://github.com/entwicklerstube/babel-plugin-root-import)) so that needs to be switched over (or de-magicked so we don't need to maintain two different babel files)
- [ ] `eslint-config-peerio` needs to be updated, all local eslint configs probably need to be tweaked. (eslint has been updated locally in this repo, so it currently works -- those changes just need to be brought over to shared config.)
- [ ] every consumer of the library needs to be changed over too
---

#### Testing instructions

I just jumped into this, but we might actually want to do this as a multi-step process:

- First, upgrade to Babel 7 across the codebase, re-run tests, make sure everything works everywhere, maybe compare the output a little by hand if we're feeling paranoid/aren't totally confident about our coverage...
- Then introduce `@babel/plugin-typescript`, change over ESLint, etc. This step is a lot safer than the Babel 7 upgrade in that there's very little risk of breakage or accidental semantic changes. But it involves more work to get everything running, so it's potentially less confusing to do separately.

----
### Repository owner


[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)
